### PR TITLE
Adds specification of default query and response handling

### PR DESCRIPTION
**Pull request by matsduf**
## Purpose

This PR adds a specification for the default setting of the DNS queries that the test cases emit. Test cases are to refer to this specification and use the default setting or specify the deviation.

1. By having a shared specification the test cases will be more consistent.
2. The test case specifications do not have to specify the DNS query when default values are used.

The specification also has default handling of the response, and that will give consistent handling in the test cases.

### CODE REVIEW

**Comment by matsduf** [comment]
@vlevigneron, do you have a comment on this PR?

File: docs/specifications/tests/DNSQueryAndResponsSpecification.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+
+## Default setting in *EDNS Query*
+
+An *EDNS Query* inherit the default setting from a *DNS Query* except for the
+setting items specified below.
+
+A test case specification can refer to an *EDNS Query* with one or several
+changes to the setting items. Specifically, the test case specification must
+always specify query name and query type.
```
Comment (mattias-p) [inline docs/specifications/tests/DNSQueryAndResponsSpecification.md:59]: Does it really make sense to mention query name and query type again here?

There's another one like this below.

File: docs/specifications/tests/DNSQueryAndResponsSpecification.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+|QR flag      | unset     | Always unset in a query        |
+|AA flag      | unset     |                                |
+|TC flag      | unset     |                                |
+|RD flag      | unset     |                                |
+|RA flag      | unset     |                                |
+|AD flag      | unset     |                                |
+|CD flag      | unset     |                                |
+|RCODE        | "NoError" | Always "NoError" in a query    |
+|Query name   | -         | Defined in test case           |
+|Query type   | -         | Defined in test case           |
```
Comment (mattias-p) [inline docs/specifications/tests/DNSQueryAndResponsSpecification.md:48]: Maybe also list query class?

This goes for the response handling too.

File: docs/specifications/tests/DNSQueryAndResponsSpecification.md [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+|Query type    | ignore             |                                 |
+|ENDS          | ignore             |                                 |
+
+
+* Check against query name and query type is, by default, done against the values
+  in the query, not in the response.
+  
+* When fetching records from the answer section, these are the default criteria:
+  * Only records matching the query name and the query type are fetched. RRSIG
+    records, when available, meeting the next criterium are also fetched.
```
Comment (mattias-p) [inline docs/specifications/tests/DNSQueryAndResponsSpecification.md:112]: Maybe remove the "when available" and the commas surrounding it? It makes the sentence more complex without helping understanding.

File: docs/specifications/tests/DNSQueryAndResponsSpecification.md [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+  * The first CNAME in the chain must match the query name.
+  * The last record in the chain must either be a CNAME or a record matching the
+    query type.
+
+* Authority and additional sections are by default ignored.
+
+
+## Default handling of an *EDNS Response*
+
+A *EDNS response* is a response to an *EDNS Query*. An *EDNS Response* inherits
```
Comment (mattias-p) [inline docs/specifications/tests/DNSQueryAndResponsSpecification.md:128]: `A` -> `An`

File: docs/specifications/tests/DNSQueryAndResponsSpecification.md [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+the default handling from a *DNS Response* except for the response items
+specified below. Unless specified in the test case specification, the items in
+the response is handled as the default handling.
```
Comment (mattias-p) [inline docs/specifications/tests/DNSQueryAndResponsSpecification.md:131]: `as` -> `using`

There's another one like this below.

**Review by mattias-p — COMMENTED** [review]
I really like this document!

The filename is misspelled. Also, maybe call it `DNSQueryAndResponseDefaults.md`?

File: docs/specifications/tests/DNSQueryAndResponsSpecification.md (hunk 3 above)
Comment (matsduf) [reply docs/specifications/tests/DNSQueryAndResponsSpecification.md:112]: Changed by putting an "Any" first in the sentence.

File: docs/specifications/tests/DNSQueryAndResponsSpecification.md (hunk 4 above)
Comment (matsduf) [reply docs/specifications/tests/DNSQueryAndResponsSpecification.md:128]: Fixed.

File: docs/specifications/tests/DNSQueryAndResponsSpecification.md (hunk 5 above)
Comment (matsduf) [reply docs/specifications/tests/DNSQueryAndResponsSpecification.md:131]: Fixed.

**Review by matsduf — COMMENTED** [review]
The document is renamed as suggested. Also the headline is changed accordingly.

**Review by mattias-p — APPROVED** [review]
LGTM

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+|Setting Item |Default value |Comment                         |
+|:------------|:-------------|:-------------------------------|
+|Protocol     | UDP          |                                |
+|OpCode       | "query"      | Always "query" in a query      |
+|QR flag      | unset        | Always unset in a query        |
+|AA flag      | unset        |                                |
+|TC flag      | unset        |                                |
+|RD flag      | unset        |                                |
+|RA flag      | unset        |                                |
```
Comment (mattias-p) [inline docs/specifications/tests/DNSQueryAndResponseDefaults.md:43]: Do we ever set the RA flag in queries?

Comment (matsduf) [reply docs/specifications/tests/DNSQueryAndResponseDefaults.md:43]: No. It has no meaning in a query, as far as I know. Possibly it could be set to see the behavior of a name server.

**Comment by matsduf** [comment]
@mattias-p, updated after our discussion. Please review again.

**Comment by matsduf** [comment]
@mattias-p, @vlevigneron, @pnax and @tgreenx, please review this.

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+changes to the setting items.
+
+|Setting Item          |Default value        |Fixed |Comment |
+|:---------------------|:--------------------|:-----|:-------|
+|EDNS                  | OPT record included | yes  |        |
+|EDNS UDP Message size | 512                 |      |        |
+|EDNS Extended RCODE   | no                  |      |        |
+|EDNS Version          | 0                   |      |        |
+|EDNS DO flag          | unset               |      |        |
+|EDNS Z lag            | unset               |      |        |
```
Comment (pnax) [inline docs/specifications/tests/DNSQueryAndResponseDefaults.md:69]: typo: lag -> flag

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+|:---------------------|:-------------|:-----|:-------|
+|EDNS DO flag          | set          | yes  |        |
+
+
+## Default handling of a *DNS Response*
+
+A *DNS Response* is a response to a *DNS Query*. Unless specified in the test
+case specification, the items in the response is handled as listed. If a
+*Response Item* is specified as "fixed" then the requirement in default action
+must be successful for the response to be considered to be a valid DNS response.
```
Comment (pnax) [inline docs/specifications/tests/DNSQueryAndResponseDefaults.md:92]: suggestion: could you  rephrase the last sentence? (using `default action` seems confusing because it does not refer to any part of the following array)

> If a *Response Item* is specified as "fixed" then the default handling must be successful for the response to be considered to be a valid DNS response.

**Review by pnax — COMMENTED** [review]
This looks good and would definitely remove some redundancy across tests specifications.

Just a question on the handling of an EDNS or DNSSEC response. If the OPT flag or the DO flag are missing, the default handling says that a warning should be raised. However the comment suggest this warning could be downgraded to a DEBUG message.
Are responses without the OPT flag or DO flag valid responses? If so why raising a warning?
Or if they are invalid, should we only display a DEBUG message?
Or am I missing something else?

**Comment by matsduf** [comment]
> Just a question on the handling of an EDNS or DNSSEC response. If the OPT flag or the DO flag are missing, the default handling says that a warning should be raised. However the comment suggest this warning could be downgraded to a DEBUG message.

Bad wording. I have rephrased it. Please re-reveiw.

> Are responses without the OPT flag or DO flag valid responses? If so why raising a warning? Or if they are invalid, should we only display a DEBUG message? Or am I missing something else?

Not really. Either the OPT should be included, or FORMERR is expected. The handling should be set in the test case specification, but at least a debug message should be issued.

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md (hunk 7 above)
Comment (matsduf) [reply docs/specifications/tests/DNSQueryAndResponseDefaults.md]: Fixed.

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md (hunk 8 above)
Comment (matsduf) [reply docs/specifications/tests/DNSQueryAndResponseDefaults.md]: Rephrased. Please re-review.

**Review by pnax — APPROVED** [review]
Thanks for the explanation. Then indeed it seems reasonable to let the test case implementation deal with the missing flag.

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+|AA flag      | unset        | yes  |                              |
+|TC flag      | unset        | yes  |                              |
+|RD flag      | unset        |      |                              |
+|RA flag      | unset        | yes  |                              |
+|AD flag      | unset        |      |                              |
+|CD flag      | unset        |      |                              |
+|RCODE        | "NoError"    | yes  |                              |
+|Query name   | -            |      | Must be defined in test case |
+|Query type   | -            |      | Must be defined in test case |
+|Query class  | "IN"         | yes  |                              |
```
Comment (vlevigneron) [inline docs/specifications/tests/DNSQueryAndResponseDefaults.md:49]: Perhaps that "class"  should not declare as a "fixed" value. Although we use "IN" in the current TC specified, I think we should not "fix" it for future usage.

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+|QR flag      | unset        | yes  | Always unset in a query      |
+|AA flag      | unset        | yes  |                              |
+|TC flag      | unset        | yes  |                              |
+|RD flag      | unset        |      |                              |
+|RA flag      | unset        | yes  |                              |
+|AD flag      | unset        |      |                              |
+|CD flag      | unset        |      |                              |
+|RCODE        | "NoError"    | yes  |                              |
+|Query name   | -            |      | Must be defined in test case |
+|Query type   | -            |      | Must be defined in test case |
```
Comment (vlevigneron) [inline docs/specifications/tests/DNSQueryAndResponseDefaults.md:48]: In the current code, there is a default value "A". For readibility reason we took the habit to specify it in the implementation of TC, but it exists.

**Review by vlevigneron — CHANGES_REQUESTED** [review]

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md (hunk 9 above)
Comment (matsduf) [reply docs/specifications/tests/DNSQueryAndResponseDefaults.md]: I really do not see the use case of other class than "IN", but I have updated the specification to make "IN" the default instead of fixed. Please re-review.

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md (hunk 10 above)
Comment (matsduf) [reply docs/specifications/tests/DNSQueryAndResponseDefaults.md:48]: This document specifies the defaults etc for the specifications to be used by the implementation. I see no reason to have A as the default query type. It is better if the query type is always spelled out. The implementation can still have that as the default query type, but A queries are less common than SOA queries, I think. I guess that having A as default is a transfer of the way `dig` and `nslookup` behaves.

**Comment by matsduf** [comment]
@vlevigneron and @pnax, please re-review after my latest update (small update).

**Review by vlevigneron — APPROVED** [review]

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+
+* Authority and additional sections are by default ignored.
+
+
+## Default handling of an *EDNS Response*
+
+An *EDNS response* is a response to an *EDNS Query*. An *EDNS Response* inherits
+the default handling from a *DNS Response* except for the response items
+specified below. Unless specified in the test case specification, the items in
+the response is handled using the default handling.
```
Comment (tgreenx) [inline docs/specifications/tests/DNSQueryAndResponseDefaults.md:135]: "is handled" -> "are handled"

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+
+The test case implementation should discover a missing OPT record and take some
+action.
+
+## Default handling of a *DNSSEC response*
+
+A *DNSSEC response* is a response to an *DNSSEC Query*. A *DNSSEC Response*
+inherits the default handling from a *EDNS Response* except for the response
+items specified below. Unless specified in the test case specification, the items
+in the response is handled using the default handling.
```
Comment (tgreenx) [inline docs/specifications/tests/DNSQueryAndResponseDefaults.md:149]: "is handled" -> "are handled"

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+
+|Setting               |Default value |Fixed |Comment |
+|:---------------------|:-------------|:-----|:-------|
+|EDNS DO flag          | set          | yes  |        |
+
+
+## Default handling of a *DNS Response*
+
+A *DNS Response* is a response to a *DNS Query*. Unless specified in the test
+case specification, the items in the response is handled as listed. If a
```
Comment (tgreenx) [inline docs/specifications/tests/DNSQueryAndResponseDefaults.md:90]: "is handled" -> "are handled"

**Review by tgreenx — DISMISSED** [review]

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+## Default handling of an *EDNS Response*
+
+An *EDNS response* is a response to an *EDNS Query*. An *EDNS Response* inherits
+the default handling from a *DNS Response* except for the response items
+specified below. Unless specified in the test case specification, the items in
+the response is handled using the default handling.
+
+|Response Item |Default handling                   | Comment                                                            |
+|:-------------|:----------------------------------|:-------------------------------------------------------------------|
+|EDNS          | Issue a message if OPT is missing | Set to DEBUG level unless specified in the test case specification |
```
Comment (mattias-p) [inline docs/specifications/tests/DNSQueryAndResponseDefaults.md:139]: Do we specify the values for the tag, testcase and module attributes of these messages anywhere? I believe it would be better if we gathered up these kinds of checks into a special kind of test case with a proper name. If you like I can create an issue for that so we don't have to deal with it now?

**Review by mattias-p — COMMENTED** [review]
I did a closer reading of the initial sections and the response handling sections and added comments for some things I found noteworthy.

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md (hunk 14 above)
Comment (matsduf) [reply docs/specifications/tests/DNSQueryAndResponseDefaults.md]: The idea was to have something in the "logs" if the OPT record was missing against expectation,  but I realize that it is probably to go beyond the purpose of this document. I have updated and removed the "automatic" debug message.

Yes, please create an issue.

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md (hunk 12 above)
Comment (matsduf) [reply docs/specifications/tests/DNSQueryAndResponseDefaults.md]: Fixed.

**Comment by matsduf** [comment]
@mattias-p, @vlevigneron  and @tgreenx, some editorial updates have been incorporated after review comments. Please re-review.

If you think that your comment has been resolved, please mark it as resolved.

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+  in the query, not in the response.
+  
+* When fetching records from the answer section, these are the default criteria:
+  * Only records matching the query name and the query type are fetched. Any
+    RRSIG records meeting the next criterium are also fetched.
+  * RRSIG records with the same query name and covering the query type are
+    fetched.
+  * CNAME records are ignored unless the query type is CNAME.
+
+* When the test case specification states that CNAME chains are to followed in
```
Comment (mattias-p) [inline docs/specifications/tests/DNSQueryAndResponseDefaults.md:121]: `to followed` -> `to be followed`

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+    RRSIG records meeting the next criterium are also fetched.
+  * RRSIG records with the same query name and covering the query type are
+    fetched.
+  * CNAME records are ignored unless the query type is CNAME.
+
+* When CNAME chains are followed (by specification) and fetch from the answer
+  section, these are the default criteria:
+  * The first CNAME in the chain must match the query name.
+  * The last record in the chain must either be a CNAME or a record matching the
+    query type.
```
[the remark this answers is not in the document]
Comment (mattias-p) [reply docs/specifications/tests/DNSQueryAndResponseDefaults.md:125]: So what happens if a response contains a CNAME chain that's invalid? Are those CNAME records simply ignored?

And if there is a valid CNAME chain in the response but also some extra unrelated CNAME records, are the records of the valid CNAME chain to be considered by the test case and are the extra CNAME records to be ignored?

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+## Default handling of a *DNSSEC response*
+
+A *DNSSEC response* is a response to a *DNSSEC Query*. A *DNSSEC Response*
+inherits the default handling from a *EDNS Response* except for the response
+items specified below. Unless specified in the test case specification, the items
+in the response are handled using the default handling.
+
+|Response Item |Default handling                       | Comment                                                            |
+|:-------------|:------------------------------------- |:-------------------------------------------------------------------|
+| EDNS DO flag | Issue a message if DO flag is missing | Set to DEBUG level unless specified in the test case specification |
```
Comment (mattias-p) [inline docs/specifications/tests/DNSQueryAndResponseDefaults.md:151]: Maybe update this to `Take note` also?

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+This specification applies to all [test case specifications][Test Cases] that has
+an explicit reference to this document.
+
+
+## Default setting in *DNS Query*
+
+A *DNS Query* has the following default setting. A test case specification can
+refer to a *DNS Query* with one or several changes to the *Parameters*
+overriding the default values. If a *Parameter* is specified as "fixed" (with a
+"X" in that column) then the default value cannot be overidden.
```
Comment (mattias-p) [inline docs/specifications/tests/DNSQueryAndResponseDefaults.md:33]: `a "X"` -> `an "X"` (there are more of these)

**Comment by mattias-p** [comment]
>>Do we specify the values for the tag, testcase and module attributes of these messages anywhere? I believe it would be better if we gathered up these kinds of checks into a special kind of test case with a proper name. If you like I can create an issue for that so we don't have to deal with it now?
>
> The idea was to have something in the "logs" if the OPT record was missing against expectation, but I realize that it is probably to go beyond the purpose of this document. I have updated and removed the "automatic" debug message.
>
> Yes, please create an issue.

I created #1034.

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (matsduf) [reply docs/specifications/tests/DNSQueryAndResponseDefaults.md:126]: Good questions.

> So what happens if a response contains a CNAME chain that's invalid? Are those CNAME records simply ignored?

Yes, unless the test case specification says anything else.

> And if there is a valid CNAME chain in the response but also some extra unrelated CNAME records, are the records of the valid CNAME chain to be considered by the test case and are the extra CNAME records to be ignored?

Yes, unless the test case specification says something else.

I added that the there must only be one CNAME record per owner name.

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md (hunk 17 above)
Comment (matsduf) [reply docs/specifications/tests/DNSQueryAndResponseDefaults.md]: Fixed.

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+* When fetching records from the answer section, these are the default criteria:
+  * Only records matching the query name and the query type are fetched. Any
+    RRSIG records meeting the next criterium are also fetched.
+  * RRSIG records with the same query name and covering the query type are
+    fetched.
+  * CNAME records are ignored unless the query type is CNAME.
+
+* When the test case specification states that CNAME chains are to be followed in
+  the the answer section, these are the default criteria for the chain to be
+  fetched:
```
Comment (mattias-p) [inline docs/specifications/tests/DNSQueryAndResponseDefaults.md:123]: Since there are two defaults in play here I suggest we make that clear from the text.

```suggestion
* When the test case specification states that CNAME chains are to be followed,
  the default handling is to look for a valid CNAME chain in the answer section
  and fetch all CNAME records in that chain.
  These are the default criteria for a CNAME chain to be valid:
```

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+|AD flag       | ignore                                   |       |                      |
+|CD flag       | ignore                                   |       |                      |
+|RCODE         | -                                        |       | Defined in test case |
+|Query name    | ignore                                   |       |                      |
+|Query type    | ignore                                   |       |                      |
+|Query class   | Require value to be same as in the query |       | Normally "IN"        |
+|EDNS          | ignore                                   |       |                      |
+
+* Check against query name and query type is, by default, done against the values
+  in the query section in the query, not in the response.
```
Comment (mattias-p) [inline docs/specifications/tests/DNSQueryAndResponseDefaults.md:112]:
```suggestion
  in the question section in the query, not in the response.
```

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+  * Only records matching the query name and the query type are fetched. Any
+    RRSIG records meeting the next criterium are also fetched.
+  * RRSIG records with the same query name and covering the query type are
+    fetched.
+  * CNAME records are ignored unless the query type is CNAME.
+
+* When the test case specification states that CNAME chains are to be followed in
+  the the answer section, these are the default criteria for the chain to be
+  fetched:
+  * The first CNAME in the chain must match the query name.
```
Comment (mattias-p) [inline docs/specifications/tests/DNSQueryAndResponseDefaults.md:124]: Is any CNAME that matches the query name by definition the first one in a chain, or does "first" refer to the order in which the records are laid out in the answer section? This should be clear from the text.

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+## Default handling of an *EDNS Response*
+
+An *EDNS response* is a response to an *EDNS Query*. An *EDNS Response* inherits
+the default handling from a *DNS Response* except for the response items
+specified below. Unless specified in the test case specification, the items in
+the response are handled using the default handling.
+
+|Response Item |Default handling             | Comment                                                        |
+|:-------------|:----------------------------|:----------------------- ---------------------------------------|
+|EDNS          | Take note if OPT is missing | Further actions to be specified in the test case specification |
```
Comment (mattias-p) [inline docs/specifications/tests/DNSQueryAndResponseDefaults.md:143]: There's something wrong with the formatting here.

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md (hunk 21 above)
Comment (matsduf) [reply docs/specifications/tests/DNSQueryAndResponseDefaults.md:124]: > Is any CNAME that matches the query name by definition the first one in a chain, 

The first CNAME record in a chain is one that no other CNAME record's RDATA points to. The last CNAME record i the chain is one whose RDATA does not point to a CNAME record (but possibly to a non-CNAME record).

> or does "first" refer to the order in which the records are laid out in the answer section?

Records are unordered in the answer section, at least no special order is required by RFC 1035. (The only RR type that requires order, as far as I know, are TSIG records that must be last in the additional section.)

> This should be clear from the text.

Or assumed to be known? I have update the text somewhat.

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md (hunk 22 above)
Comment (matsduf) [reply docs/specifications/tests/DNSQueryAndResponseDefaults.md:143]: Fixed.

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md [hunk 23]
```diff
@@ -146,12 +150,10 @@ inherits the default handling from a *EDNS Response* except for the response
 items specified below. Unless specified in the test case specification, the items
 in the response are handled using the default handling.
 
-|Response Item |Default handling                       | Comment                                                            |
-|:-------------|:------------------------------------- |:-------------------------------------------------------------------|
-| EDNS DO flag | Issue a message if DO flag is missing | Set to DEBUG level unless specified in the test case specification |
+|Response Item |Default handlin                  | Comment                                                        |
```
[the remark this answers is not in the document]
Comment (matsduf) [reply docs/specifications/tests/DNSQueryAndResponseDefaults.md:153]: Fixed.

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md [hunk 24]
```diff
@@ -0,0 +1,165 @@
...
+|RA flag       | ignore                                   |       |                      |
+|AD flag       | ignore                                   |       |                      |
+|CD flag       | ignore                                   |       |                      |
+|RCODE         | -                                        |       | Defined in test case |
+|Query name    | ignore                                   |       |                      |
+|Query type    | ignore                                   |       |                      |
+|Query class   | Require value to be same as in the query |       | Normally "IN"        |
+|EDNS          | ignore                                   |       |                      |
+
+* Check against query name and query type is, by default, done against the values
```
Comment (tgreenx) [inline docs/specifications/tests/DNSQueryAndResponseDefaults.md:112]: For better readability, maybe change the "response item" values (i.e., query name, query type, etc) in the text below to be in italic for example? Or double quoted.

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md [hunk 25]
```diff
@@ -0,0 +1,165 @@
...
+  in the question section in the query, not in the response.
+  
+* When fetching records from the answer section, these are the default criteria:
+  * Only records matching the query name and the query type are fetched. Any
+    RRSIG records meeting the next criterium are also fetched.
```
Comment (tgreenx) [inline docs/specifications/tests/DNSQueryAndResponseDefaults.md:117]: Criterium -> Criterion

**Review by tgreenx — APPROVED** [review]

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md (hunk 24 above)
Comment (matsduf) [reply docs/specifications/tests/DNSQueryAndResponseDefaults.md:112]: Updated.

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md (hunk 25 above)
Comment (matsduf) [reply docs/specifications/tests/DNSQueryAndResponseDefaults.md:117]: Updated.

**Comment by matsduf** [comment]
@mattias-p, @pnax and @mattias-p, please re-review.

**Review by tgreenx — APPROVED** [review]

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md [hunk 26]
```diff
@@ -0,0 +1,165 @@
...
+  the default handling is to look for a valid CNAME chain in the answer section
+  and fetch all records in that chain. The records in the chain are arranged in
+  the logical order. These are the default criteria for a CNAME chain to be
+  valid:
+  * The first CNAME in the chain must match *Query Name*.
+  * The last record in the chain must either be a CNAME or a record matching
+    *Query Type*.
+  * For each owner name of the CNAME records in the chain there must not be any
+    additional CNAME records in the answer section (only one CNAME record per
+    owner name).
```
Comment (mattias-p) [inline docs/specifications/tests/DNSQueryAndResponseDefaults.md:132]: I'm curious about an edge case here. Let's say we're following a CNAME chain and the query name is example.com. The answer section contains the following CNAME records:
```
error.com    3600  IN  CNAME  example.com
example.com  3600  IN  CNAME  other.com
```
The way I read the specification there's one chain in the answer section, and the first CNAME is error.com, which is different from the query name, so none of these records should be fetched. Is that the intended behavior?

I believe the "The first CNAME in the chain must match Query Name." item should be removed from the list of criteria and added as a definition instead.

Comment (matsduf) [reply docs/specifications/tests/DNSQueryAndResponseDefaults.md:132]: > I'm curious about an edge case here. Let's say we're following a CNAME chain and the query name is example.com. The answer section contains the following CNAME records:
> 
> ```
> error.com    3600  IN  CNAME  example.com
> example.com  3600  IN  CNAME  other.com
> ```
> 
> The way I read the specification there's one chain in the answer section, and the first CNAME is error.com, which is different from the query name, so none of these records should be fetched. Is that the intended behavior?

Yes.

> I believe the "The first CNAME in the chain must match Query Name." item should be removed from the list of criteria and added as a definition instead.

I disagree, but I will change the wording somewhat to make intended behavior clear.

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md [hunk 27]
```diff
@@ -28,40 +28,40 @@ an explicit reference to this document.
...
-|Setting Item |Default value |Fixed |Comment                       |
+|Parameter    |Default value |Fixed |Comment                       |
 |:------------|:-------------|:-----|:-----------------------------|
 |Protocol     | UDP          |      |                              |
-|OpCode       | "query"      | yes  |                              |
-|QR flag      | unset        | yes  | Always unset in a query      |
-|AA flag      | unset        | yes  |                              |
-|TC flag      | unset        | yes  |                              |
+|OpCode       | "query"      | X    |                              |
+|QR flag      | unset        | X    | Always unset in a query      |
```
Comment (pnax) [inline docs/specifications/tests/DNSQueryAndResponseDefaults.md:39]: Why specify here in the comment that the parameter is unset since it is already fixed to `unset`?

**Comment by matsduf** [comment]
@mattias-p, I have updated based on your comment. Please re-review.

@tgreenx, I have done som updates based on comments from @mattias-p. Please re-review.

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md (hunk 27 above)
Comment (matsduf) [reply docs/specifications/tests/DNSQueryAndResponseDefaults.md:39]: To underline that fact.

**Comment by matsduf** [comment]
@pnax, please re-review.

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md [hunk 28]
```diff
@@ -120,19 +120,27 @@ to be considered to be a DNS response.
...
+    * The last record in the chain must either be a CNAME or a record matching
+      *Query Type*.
+    * For each owner name of the CNAME records in the chain there must not be any
+      additional records in the answer section with the same owner name besides
+      RRSIG and NSEC records (that may exist).
+  * CNAME records not part of a valid CNAME chain are by default ignored.
 
 * Authority and additional sections are by default ignored.
 
+The test case specification may prescribe that CNAME records are handled in an a
```
Comment (pnax) [inline docs/specifications/tests/DNSQueryAndResponseDefaults.md:141]: typo: `an a` -> `a`

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md [hunk 29]
```diff
@@ -120,19 +120,27 @@ to be considered to be a DNS response.
...
-    *Query Type*.
-  * For each owner name of the CNAME records in the chain there must not be any
-    additional CNAME records in the answer section (only one CNAME record per
-    owner name).
+  the default handling is to only follow a CNAME, and fetch the records, if the
+  CNAME chain is valid.
+  * The chain is, by default, considered to be valid if the following criteria
+    are met:
+    * It must be possible to arrange all CNAME records from the answer section
+      into a contiguous logical chain with a posisble addition of a non-CNAME
```
Comment (pnax) [inline docs/specifications/tests/DNSQueryAndResponseDefaults.md:128]: typo: `posisble` -> `possible`

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md (hunk 27 above)
Comment (matsduf) [reply docs/specifications/tests/DNSQueryAndResponseDefaults.md:39]: It is just a comment. The choice of the other fixed, and default, values could also be commented, but I have not done that.

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md (hunk 28 above)
Comment (matsduf) [reply docs/specifications/tests/DNSQueryAndResponseDefaults.md]: Fixed.

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md (hunk 27 above)
Comment (matsduf) [reply docs/specifications/tests/DNSQueryAndResponseDefaults.md]: Removed.

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md (hunk 29 above)
Comment (matsduf) [reply docs/specifications/tests/DNSQueryAndResponseDefaults.md]: Fixed.

**Comment by matsduf** [comment]
@pnax, please re-review.

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md (hunk 24 above)
Comment (tgreenx) [reply docs/specifications/tests/DNSQueryAndResponseDefaults.md:112]: Missing italic for "query name" and "query type" on line 112.

**Review by tgreenx — DISMISSED** [review]

File: docs/specifications/tests/DNSQueryAndResponseDefaults.md (hunk 24 above)
Comment (matsduf) [reply docs/specifications/tests/DNSQueryAndResponseDefaults.md:112]: When I read this again, I feel I have to check the wording of this sentence.

Comment (matsduf) [reply docs/specifications/tests/DNSQueryAndResponseDefaults.md]: Updated wording.

**Comment by matsduf** [comment]
@tgreenx, please re-review.

**Review by mattias-p — APPROVED** [review]
Awesome!

**Review by tgreenx — APPROVED** [review]
